### PR TITLE
CloudFront Signed Url and Cookie (#3529)

### PR DESCRIPTION
**Pull request by davidh44**
Merging CloudFront Signed URLs and Cookies into master

## Motivation and Context

Closes feature parity from v1

https://github.com/aws/aws-sdk-java-v2/issues/370

## Modifications

Added classes and tests

## Testing

Added unit and integration tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/cloudfront/pom.xml [hunk 1]
```diff
@@ -56,5 +56,15 @@
...
         </dependency>
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>s3</artifactId>
+            <version>${awsjavasdk.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>nl.jqno.equalsverifier</groupId>
+            <artifactId>equalsverifier</artifactId>
```
Comment (zoewangg) [inline services/cloudfront/pom.xml:67]: Should be test scope

File: services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        return getSignedUrlWithCannedPolicy(CannedSignerRequest.builder().applyMutation(request).build());
+    }
+
+    /**
+     * Returns a signed URL with a canned policy that grants universal access to
+     * private content until a given date.
+     * For more information, see <a href=
+     * "https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-canned-policy.html"
+     * >Creating a signed URL using a canned policy</a>.
+     *
```
Comment (zoewangg) [inline services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java:95]: Can we add some code snippet for all public APIs?

> Use @snippet to add code snippets.

https://github.com/aws/aws-sdk-java-v2/blob/master/docs/design/APIReference.md#style-guideline

File: services/cloudfront/pom.xml (hunk 1 above)
Comment (davidh44) [reply services/cloudfront/pom.xml:67]: updated to test scope

File: services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+import software.amazon.awssdk.services.cloudfront.internal.url.DefaultSignedUrl;
+import software.amazon.awssdk.services.cloudfront.internal.utils.SigningUtils;
+import software.amazon.awssdk.services.cloudfront.model.CannedSignerRequest;
+import software.amazon.awssdk.services.cloudfront.model.CustomSignerRequest;
+import software.amazon.awssdk.services.cloudfront.url.SignedUrl;
+
+/**
+ *
+ * Utilities for working with CloudFront distributions
+ * <p>
```
[the remark this answers is not in the document]
Comment (davidh44) [reply services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java:40]: removed extra <p>

File: services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/internal/auth/DerParser.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    public DerParser(InputStream in) {
+        this.in = in;
+    }
+
+    /**
+     * Create a new DER decoder from a byte array.
+     *
+     * @param the encoded bytes
+     */
+    public DerParser(byte[] bytes) {
```
Comment (dave-fn) [inline services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/internal/auth/DerParser.java]: nit: missing param name

File: services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java (hunk 2 above)
Comment (davidh44) [reply services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java:92]: to add inline code snippets

File: services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/internal/cookie/DefaultCookiesForCannedPolicy.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return "CloudFront Cookies for Canned Policy:\n"
+               + "Resource URL = " + resourceUrl + "\n"
+               + "Signature Header Value = " + signatureHeaderValue + "\n"
+               + "Key-Pair-Id Header Value = " + keyPairIdHeaderValue + "\n"
+               + "Expires Header Value = " + expiresHeaderValue;
+    }
```
Comment (dave-fn) [inline services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/internal/cookie/DefaultCookiesForCannedPolicy.java:60]: Might be easier to use `ToString.builder()`?

Comment (davidh44) [reply services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/internal/cookie/DefaultCookiesForCannedPolicy.java:60]: updated to use ToString

**Review by zoewangg — APPROVED** [review]

**Pull request merged**